### PR TITLE
Perf(LTX2): Comprehensive XLA, Memory, and Transformer Code Quality Optimizations

### PR DESCRIPTION
**Pull request by Perseus14**
## LTX2 Improvements, Optimizations & Bug Fixes

This PR introduces a massive overhaul to the LTX2 and LTX2.3 implementations in MaxDiffusion. The primary focus spans architectural cleanliness, persistent memory caching, AOT loop optimizations, layout sharding stability, distributed attention compatibility, and dramatic improvements to end-to-end latency (both compilation and post-processing).

### 🌟 Architectural & JAX Optimizations
* **Dataclass Contexts**: Replaced dictionary-based contexts with `flax.struct.dataclass` (`LTX2BlockContext`), significantly improving code readability and type safety across attention mechanisms.
* **Dynamic Float Scales**: Shifted floating-point parameters (`guidance_scale`, `stg_scale`) out of `static_argnames` and introduced boolean triggers (`do_cfg`, `do_stg`). This drastically minimizes recompilation overhead and tracing stalls.
* **Guidance Path Deduplication**: Refactored `_apply_ltx2_guidance` to heavily deduplicate and unify the computational paths for standard CFG, STG, and audio guidance logic, resulting in a significantly cleaner XLA graph.

### ⚡ Caching, Loading, and Compilation (Ports of PRs #438 & #443)
* **Cross-Attention KV Cache Introduced**: Fully implemented and wired the Cross-Attention KV Cache system for LTX2. By caching the text and audio embedding projections, we eliminate redundant cross-attention compute during every step of the diffusion loop. Added strict checks to reject incompatible caches and properly wired named scopes.
* **Persistent Converted-Weight Caching**: Introduced a robust, WAN-style atomic persistent memoization system (`try_load_converted_weights`) for LTX2. This safely caches converted weights to disk, drastically accelerating model loading on subsequent runs by bypassing redundant conversions.
* **Fast Safetensor Weight Loading**: Eliminated memory-heavy `torch.load` and sequential `torch2jax` routines in favor of mapping `.safetensors` concurrently with a `ThreadPoolExecutor`. Memory ingestion is aggressively parallelized using zero-copy reinterpretation (`tensor.view(torch.uint16).numpy().view(ml_dtypes.bfloat16)`).
* **AOT Diffusion Caching & Warmup**: Swapped `@jax.jit` for `@aot_cache.cached_jit` on the main diffusion loop. Integrated 2-step fast-warmup logic (`with aot_cache.warmup_mode()`) which ensures the full step count is traced when `scan_diffusion_loop=True`, bypassing prolonged compilations on application startup.
* **ICI-Staged Device Put**: Large tensors (>=64MB) are now pushed directly to device 0 and replicated over ICI (when `jax.process_count() == 1`), rather than enduring slow host-to-device PCIe copies for all devices.

### 🚀 Post-Processing Latency
* **Zero-Copy NumPy Export**: Introduced the `np_uint8` pipeline output type and updated the YAML configs to default to it. This entirely bypasses the massive CPU bottleneck of sequentially instantiating `PIL.Image` objects across 121 frames. 
* **Native TPU Vectorized Normalization**: Completely bypassed the standard PyTorch/NumPy CPU-based video processor. By executing the tensor normalization (clipping, scaling, and casting to `uint8` or `float32`) as vectorized JAX operations natively on the TPU *prior* to host transfer, we eliminate massive host-side memory thrashing and CPU iteration overhead.

### 🌀 Distributed Attention
* **Ulysses & Ring Attention Support**: Fully wired `ring_attention` and `ulysses_attention` parameters down through all `LTX2VideoTransformerBlock` instantiations (affecting both `nnx.scan` and un-scanned eager block initializations).
* **Cross-Attention Axis Fallbacks**: Implemented dynamic axis-mapping for `NNXAttentionOp` in LTX2 cross-attention mechanisms, ensuring layout rules bind properly when Ulysses sharding configurations are enabled.
* **Tile-Size Auto-Tuning**: Integrated LTX2/LTX2.3 enhancements specifically targeting tile-size auto-tuning for spatial operations.

### 📊 Results

**DiT Execution Time:**
- **v6e-8**: `11.2s` ➔ `9.5s`
- **7x-8**: `7.8s` ➔ `5.6s`

**Video Post-Processing Time:**
- **v6e-8**: `0.200s` ➔ `0.040s`
- **7x-8**: `0.500s` ➔ `0.052s`

### CODE REVIEW

File: src/maxdiffusion/models/ltx2/text_encoders/torchax_text_encoder.py [hunk 1]
```diff
@@ -57,8 +55,7 @@ def __call__(
       self, input_ids: jax.Array, attention_mask: jax.Array, output_hidden_states: bool = True
   ) -> Tuple[jax.Array, ...]:
     # Dynamically patch transformers.masking_utils only during the duration of this call
-    transformers.masking_utils.sliding_window_overlay = _patched_sliding_window_overlay
-    try:
+    with mock.patch("transformers.masking_utils.sliding_window_overlay", _patched_sliding_window_overlay):
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/models/ltx2/text_encoders/torchax_text_encoder.py]: Done!

File: src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py [hunk 2]
```diff
@@ -1555,12 +1554,22 @@ def __call__(
...
+        @jax.jit
+        def enforce_layout_act(x):
+          return jax.lax.with_sharding_constraint(x, activation_axes)
+
+        @jax.jit
+        def enforce_layout_act_audio(x):
+          return jax.lax.with_sharding_constraint(x, activation_axes_audio)
+
+        latents_jax = enforce_layout_act(latents_jax)
+        audio_latents_jax = enforce_layout_act_audio(audio_latents_jax)
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py]: Done!

File: src/maxdiffusion/generate_ltx2.py [hunk 3]
```diff
@@ -237,7 +237,8 @@ def run(config, pipeline=None, filename_prefix="", commit_hash=None):
 
   # Export videos
   for i in range(len(videos)):
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/generate_ltx2.py:414]: Done!

File: src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py [hunk 4]
```diff
@@ -81,6 +81,8 @@ def rescale_noise_cfg(noise_cfg, noise_pred_text, guidance_rescale=0.0):
   """
   std_text = jnp.std(noise_pred_text, axis=list(range(1, noise_pred_text.ndim)), keepdims=True)
   std_cfg = jnp.std(noise_cfg, axis=list(range(1, noise_cfg.ndim)), keepdims=True)
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py:138]: Done!

**Comment by Perseus14** [comment]
@mbohlool Can you please review this?

File: src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py [hunk 5]
```diff
@@ -1555,12 +1559,14 @@ def __call__(
       audio_embeds_sharded = audio_embeds
 
       if not self.transformer.scan_layers:
-        activation_axes = nn.logical_to_mesh_axes(("activation_batch", "activation_length", "activation_embed"))
-        activation_axes_audio = nn.logical_to_mesh_axes(("activation_batch", None, "activation_embed"))
-        spec = NamedSharding(self.mesh, P(*activation_axes))
-        spec_audio = NamedSharding(self.mesh, P(*activation_axes_audio))
-        video_embeds_sharded = jax.device_put(video_embeds, spec)
-        audio_embeds_sharded = jax.device_put(audio_embeds, spec_audio)
+        with nn_partitioning.axis_rules(self.config.logical_axis_rules):
```
Comment (entrpn) [inline src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py:2032]: no `with mesh` required here?

**Review by entrpn — DISMISSED** [review]

File: src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py (hunk 5 above)
Comment (Perseus14) [reply src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py:2032]: Thanks for reviewing. I am using `_enforce_layout` which is decorated with @jax.jit and looks like it auto infers the mesh based on the shape of the tensors.

File: src/maxdiffusion/models/ltx2/transformer_ltx2.py [hunk 6]
```diff
@@ -1195,38 +1272,47 @@ def __call__(
...
+          video_rotary_emb=video_rotary_emb,
+          audio_rotary_emb=audio_rotary_emb,
+          ca_video_rotary_emb=video_cross_attn_rotary_emb,
+          ca_audio_rotary_emb=audio_cross_attn_rotary_emb,
+          encoder_attention_mask=encoder_attention_mask,
+          audio_encoder_attention_mask=audio_encoder_attention_mask,
+          modality_mask=modality_mask,
+      )
+
+      def apply_block(block, context: LTX2BlockContext, mask, rngs_carry):
```
Comment (prishajain1) [inline src/maxdiffusion/models/ltx2/transformer_ltx2.py]: rngs_carry seems to be passed unchanged. Do we need to pass it here?

File: src/maxdiffusion/models/ltx2/transformer_ltx2.py [hunk 7]
```diff
@@ -1195,38 +1272,47 @@ def __call__(
         audio_encoder_hidden_states = audio_encoder_hidden_states.reshape(batch_size, -1, audio_hidden_states.shape[-1])
     # 5. Run transformer blocks
     with jax.named_scope("Transformer Blocks"):
+      base_context = LTX2BlockContext(
+          hidden_states=hidden_states,
+          audio_hidden_states=audio_hidden_states,
+          encoder_hidden_states=encoder_hidden_states,
+          audio_encoder_hidden_states=audio_encoder_hidden_states,
```
Comment (prishajain1) [inline src/maxdiffusion/models/ltx2/transformer_ltx2.py:1605]: do we need to pass a2v and v2a cross attn masks here?

File: src/maxdiffusion/models/ltx2/transformer_ltx2.py (hunk 6 above)
Comment (Perseus14) [reply src/maxdiffusion/models/ltx2/transformer_ltx2.py]: Fixed!

File: src/maxdiffusion/models/ltx2/transformer_ltx2.py (hunk 7 above)
Comment (Perseus14) [reply src/maxdiffusion/models/ltx2/transformer_ltx2.py:1605]: Fixed!

File: src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py [hunk 8]
```diff
@@ -1202,13 +1536,35 @@ def _decode_latents_to_video(
...
+    # Fast post-processing: Format on TPU to save bandwidth, then gather and wrap
     t0_video_post = time.perf_counter()
+
+    # 1. Native TPU Formatting
+    if output_type in ["pil", "np_uint8"]:
+      # PIL and raw numpy both need (B, T, H, W, C) in uint8.
+      video = _tpu_format_pil(video)
+    elif output_type in ["np", "pt"]:
+      # PyTorch/NumPy expect (B, C, T, H, W) in float [0.0, 1.0].
+      video = _tpu_format_pt(video)
```
Comment (prishajain1) [inline src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py:1548]: nit: output dimensions check: np and pt now return (B, C, T, H, W), whereas they previously returned (B, T, H, W, C), hope that's an intended change. Doesn't affect the mp4 videos produced though.

File: src/maxdiffusion/configs/ltx2_video.yml [hunk 9]
```diff
@@ -1,7 +1,18 @@
 #hardware
 hardware: 'tpu'
 skip_jax_distributed_system: False
+# Supported attention kernels:
+# dot_product, flash, tokamax_flash, tokamax_ring, tokamax_ring_custom,
+# ulysses, ulysses_custom, ulysses_custom_fixed_m, ulysses_ring,
+# ulysses_ring_custom, ulysses_ring_custom_fixed_m, ulysses_ring_custom_bidir,
+# and cudnn_flash_te (GPU only).
 attention: 'flash'
+use_base2_exp: False
```
Comment (prishajain1) [inline src/maxdiffusion/configs/ltx2_video.yml:10]: is use_ base2_exp intentionally kept off? Does this not provide improvements here?

File: src/maxdiffusion/pyconfig.py
Comment (prishajain1) [inline src/maxdiffusion/pyconfig.py:1]: should we validate context % ulysses_shards == 0, attention head counts % ulysses_shards==0 here?

File: src/maxdiffusion/configs/ltx2_video.yml (hunk 9 above)
Comment (Perseus14) [reply src/maxdiffusion/configs/ltx2_video.yml:10]: Yes it's intentionally kept off as enabling it produces a slightly different latent. It does provide latency improvement.

File: src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py (hunk 8 above)
Comment (Perseus14) [reply src/maxdiffusion/pipelines/ltx2/ltx2_pipeline.py:1548]: Yes it is to closely align with Diffusers output. No latency or output change

File: src/maxdiffusion/pyconfig.py
Comment (Perseus14) [reply src/maxdiffusion/pyconfig.py:1]: That's a good idea. I plan to create a separate PR that cleans up al the different attention kernel defintions, will incorporate that here

**Review by prishajain1 — DISMISSED** [review]
